pkg/cmd/codespace/ssh.go

```diff
@@ -236,48 +232,73 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
+// selectSSHKeys evaluates available key pairs and select which should be used to connect to the codespace
+// using the precedence rules below. If there is no error, a keypair is always returned and additionally a
+// bool flag is returned to specify if the private key need be appended to the ssh arguments (it doesn't need
+// to be if the key was selected from a -i argument).
+//
+// Precedence rules:
+// 1. Key which is specified by -i
+// 2. Automatic key, if it already exists
+// 3. First valid keypair in ssh config (according to ssh -G)
+// 4. Automatic key, newly created
+func selectSSHKeys(
 	ctx context.Context,
 	sshContext ssh.Context,
-	apiClient apiClient,
 	args []string,
 	opts sshOptions,
-) (bool, error) {
+) (*ssh.KeyPair, bool, error) {
 	customConfigPath := ""
 	for i := 0; i < len(args); i += 1 {
 		arg := args[i]
 
 		if arg == "-i" {
-			if i+1 < len(args) && path.Base(args[i+1]) == automaticPrivateKeyName {
-				return true, nil
+			if i+1 >= len(args) {
+				return nil, false, errors.New("missing value to -i argument")
 			}
 
-			// User specified a custom identity file so just trust it is correct
-			return false, nil
+			// User manually specified an identity file so just trust it is correct
+			return &ssh.KeyPair{
+				PrivateKeyPath: args[i+1],
+				PublicKeyPath:  args[i+1] + ".pub",
+			}, false, nil
 		}
 
-		if arg == "-F" && i < len(args)-1 {
+		if arg == "-F" && i+1 < len(args) {
 			// ssh only pays attention to that last specified -F value, so it's correct to overwrite here
 			customConfigPath = args[i+1]
 		}
 	}
 
-	if automaticKeySSHKeysExist(sshContext) {
+	if autoKeyPair := automaticSSHKeyPair(sshContext); autoKeyPair != nil {
 		// If the automatic keys already exist, just use them
-		return true, nil
+		return autoKeyPair, true, nil
+	}
+
+	configKeyPair, err := firstConfiguredKeyPair(ctx, customConfigPath, opts.profile)
+	if err != nil {
+		return nil, false, fmt.Errorf("checking configured keys: %v", err)
+	}
+
+	// Even if there was no error, there may simply be no configured keys, so still check for nil here
+	if configKeyPair != nil {
+		return configKeyPair, true, nil
 	}
 
-	// If there is a public key uploaded which matches a configured
-	// private key, there is no need for automatic key generation
-	hasPublicKey, err := hasUploadedPublicKeyForConfig(ctx, apiClient, customConfigPath, opts.profile)
+	// Finally, fall back to generating new automatic keys
+	autoKeyPair, err := setupAutomaticSSHKeys(sshContext)
+	if err != nil {
+		return nil, false, fmt.Errorf("generating automatic keypair: %v", err)
+	}
 
-	return !hasPublicKey, err
+	return autoKeyPair, true, nil
 }
 
-func automaticKeySSHKeysExist(sshContext ssh.Context) bool {
+// automaticSSHKeyPair returns the paths to the automatic key pair files, if they both exist
+func automaticSSHKeyPair(sshContext ssh.Context) *ssh.KeyPair {
 	publicKeys, err := sshContext.LocalPublicKeys()
 	if err != nil {
-		return false
+		return nil
```

Comment:
Why do we not return this error?

Reply:
Really the only error that we can get there is that the `~/.ssh` directory doesn't exist, which in this case is valid and means that the automatic keypair cannot exist

Reply:
Added a comment to explain this